xmr-btc/src/tor.rs

```diff
@@ -0,0 +1,127 @@
+use anyhow::{anyhow, bail, Result};
+use std::{
+    future::Future,
+    net::{IpAddr, Ipv4Addr, SocketAddr, SocketAddrV4},
+};
+use tokio::net::TcpStream;
+use torut::{
+    control::{AsyncEvent, AuthenticatedConn, ConnError, UnauthenticatedConn},
+    onion::TorSecretKeyV3,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct UnauthenticatedConnection {
+    tor_proxy_address: SocketAddrV4,
+    tor_control_port_address: SocketAddr,
+}
+
+impl Default for UnauthenticatedConnection {
+    fn default() -> Self {
+        Self {
+            tor_proxy_address: SocketAddrV4::new(Ipv4Addr::LOCALHOST, 9050),
+            tor_control_port_address: SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, 9051)),
+        }
+    }
+}
+
+impl UnauthenticatedConnection {
+    pub fn with_ports(proxy_port: u16, control_port: u16) -> Self {
+        Self {
+            tor_proxy_address: SocketAddrV4::new(Ipv4Addr::LOCALHOST, proxy_port),
+            tor_control_port_address: SocketAddr::V4(SocketAddrV4::new(
+                Ipv4Addr::LOCALHOST,
+                control_port,
+            )),
+        }
+    }
+
+    /// checks if tor is running
+    async fn tor_running(&self) -> Result<()> {
```

Comment:
Since this returns an error (i.e., this function is not a predicate), perhaps we could call it `assert_tor_running`